### PR TITLE
MediaConsortium bid adapter: initial release

### PR DESCRIPTION
**Pull request by yki-t**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter  
- [ ] Updated bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

New bid adapter

- contact email of the adapter’s maintainer
    mediaconsortium-develop@bi.garage.co.jp
- test parameters for validating bids:
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

### CODE REVIEW

File: modules/mediaConsortiumBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    return storedOptimizations ? JSON.parse(storedOptimizations) : {}
+  } catch (err) {
+    return {}
+  }
+}
+
+function getFpIdFromLocalStorage() {
+  try {
+    return window.localStorage.getItem('ope_fpid')
```
Comment (patmmccann) [inline modules/mediaConsortiumBidAdapter.js:180]: You must use storage manager, not directly access storage

File: modules/mediaConsortiumBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        cpm,
+        currency,
+        dealId,
+        ttl,
+        netRevenue: true,
+        creativeId: id,
+        mediaType,
+        width,
+        height,
+        ad: markupWithMacroReplaced,
```
Comment (patmmccann) [inline modules/mediaConsortiumBidAdapter.js:142]: I recommend you return the price with no macro, publishers could fork and change what you pay them

File: modules/mediaConsortiumBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      return acc.concat({type: syncType, url})
+    }, [])
+  }
+}
+
+registerBidder(spec)
+
+export function getOptimizationsFromLocalStorage() {
+  try {
+    const storedOptimizations = localStorage.getItem(OPTIMIZATIONS_STORAGE_KEY)
```
Comment (patmmccann) [inline modules/mediaConsortiumBidAdapter.js:170]: This is not allowed either for the same reason

File: modules/mediaConsortiumBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  supportedMediaTypes: [BANNER],
+  isBidRequestValid(bid) {
+    return true
+  },
+  buildRequests(bidRequests, bidderRequest) {
+    const {
+      auctionId,
+      bids,
+      gdprConsent: {gdprApplies = false, consentString} = {},
+      ortb2: {device, site}
```
Comment (patmmccann) [inline modules/mediaConsortiumBidAdapter.js:29]: The user object can be useful :)

**Comment by patmmccann** [comment]
> Whoa there, partner! 🌵🤠 We wrangled some duplicated code in your PR:
> 
> * [`modules/mediaConsortiumBidAdapter.js`](https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/mediaConsortiumBidAdapter.js#L243-L256) has 16 duplicated lines with [`modules/voxBidAdapter.js`](https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/voxBidAdapter.js#L49-L62)
> 
> Reducing code duplication by importing common functions from a library not only makes our code cleaner but also easier to maintain. Please move the common code from both files into a library and import it in each. We hate that we have to mention this, however, commits designed to hide from this utility by renaming variables or reordering an object are poor conduct. We will not look upon them kindly! Keep up the great work! 🚀

Would be nice to see /libraries/apnPlayerUtils... With https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/voxBidAdapter.js#L48

File: modules/mediaConsortiumBidAdapter.js (hunk 2 above)
Comment (mlequain) [reply modules/mediaConsortiumBidAdapter.js:173]: You mean replacing the `AUCTION_PRICE` macro server-side instead of doing it here ?

**Comment by mlequain** [comment]
> > Whoa there, partner! 🌵🤠 We wrangled some duplicated code in your PR:
> > 
> > * [`modules/mediaConsortiumBidAdapter.js`](https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/mediaConsortiumBidAdapter.js#L243-L256) has 16 duplicated lines with [`modules/voxBidAdapter.js`](https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/voxBidAdapter.js#L49-L62)
> > 
> > Reducing code duplication by importing common functions from a library not only makes our code cleaner but also easier to maintain. Please move the common code from both files into a library and import it in each. We hate that we have to mention this, however, commits designed to hide from this utility by renaming variables or reordering an object are poor conduct. We will not look upon them kindly! Keep up the great work! 🚀
> 
> Would be nice to see /libraries/apnPlayerUtils... With
> 
> https://github.com/prebid/Prebid.js/blob/23cfd8642fde90f3bb853ee8e198735dc91cfb4f/modules/voxBidAdapter.js#L48

Unfortunately since the video player is closed source/documentation, I don't have all the elements needed to create such a file but maybe someone from appnexus/xandr could do it ?

File: modules/mediaConsortiumBidAdapter.js (hunk 2 above)
Comment (patmmccann) [reply modules/mediaConsortiumBidAdapter.js:176]: Yes

**Comment by Rothalack** [comment]
Hey @yki-t you'll need to get a PR up on the documentation repo for the docs site. You can find that [here](https://github.com/prebid/prebid.github.io/pulls), there's lots of examples there.
https://github.com/prebid/prebid.github.io/pulls

File: modules/mediaConsortiumBidAdapter.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            code: 'div-prebid-banner',
+            mediaTypes:{
+                banner: {
+                    sizes: [[300, 250]],
+                }
+            },
+            bids:[
+                {
+                    bidder: 'mediaConsortium',
```
Comment (Rothalack) [inline modules/mediaConsortiumBidAdapter.md:56]: You show mediaConsortium here, but define your bidder code as MediaConsortium in the adapter. When you try to run an auction on mediaConsortium, it says that bidder doesn't exist. I'd say define the bidder code as mediaConsortium.

Comment (mlequain) [reply modules/mediaConsortiumBidAdapter.md:56]: The module has been renamed.

**Comment by mlequain** [comment]
> Hey @yki-t you'll need to get a PR up on the documentation repo for the docs site. You can find that [here](https://github.com/prebid/prebid.github.io/pulls), there's lots of examples there. https://github.com/prebid/prebid.github.io/pulls

Hi, the documentation PR has been submitted and approved [here](https://github.com/prebid/prebid.github.io/pull/5546).

Let me know if you need anything else

**Review by Rothalack — APPROVED** [review]
Looks good!

**Pull request merged**